src/App.js

```diff
@@ -1,101 +1,129 @@
+import './App.css'
+
+// Defining initial state outside of the component can make the component easier to read.
+const timersInitialState = [
+  {
+    id: 1,
+    name: 'Exercise 1',
+    seconds: 20,
+    minutes: 1,
+    // I'm not going to alter the shape of this object since it is used all over the app,
+    //but I would suggest not storing anything in state that can be computed from other values in the state.
+    // For example, `isComplete`, can be calculated on the fly from `currentRep` and `targetRep`. (const isComplete = currentRep >= targetRep)
+    isComplete: false,
+    currentRep: 0,
+    targetRep: 1,
+  },
+  {
+    id: 2,
+    name: 'Exercise 2',
+    seconds: 30,
+    minutes: 0,
+    isComplete: false,
+    currentRep: 0,
+    targetRep: 2,
+  },
+  {
+    id: 3,
+    name: 'Exercise 3',
+    seconds: 5,
+    minutes: 0,
+    isComplete: false,
+    currentRep: 0,
+    targetRep: 1,
+  },
+]
 
 function App() {
-  const [timers, setTimers] = useState([
-      {
-        id: 1,
-        name: "Exercise 1",
-        seconds: 20,
-        minutes: 1,
-        isComplete: false,
-        currentRep: 0,
-        targetRep: 1,
-      },
+  const [timers, setTimers] = useState(timersInitialState)
+
+  function changeCompletion_old(id) {
+    // This function is very hard to read & follow what it is doing. Good variable naming and Array methods can help improve it.
+    // I am leaving it here for reference, but renamed it to `changeCompletion_old`
+    setTimers((prevTimer) => [
+      ...prevTimer.slice(0, id - 1), // making assumptions about how the `id` relates to a timer's index is fragile and bug prone. Use the id to find the index using prevTimer.findIndex(timer => timer.id === id)
       {
-        id: 2,
-        name: "Exercise 2",
-        seconds: 30,
-        minutes: 0,
-        isComplete: false,
-        currentRep: 0,
-        targetRep: 2
+        ...prevTimer[id - 1],
+        isComplete: prevTimer[id - 1].currentRep >= prevTimer[id - 1].targetRep - 1 && true, // it works due to - 1, but i think there's something wrong with useEffect
+        currentRep: prevTimer[id - 1].currentRep + 1,
       },
-      {
-        id: 3,
-        name: "Exercise 3",
-        seconds: 5,
-        minutes: 0,
-        isComplete: false,
-        currentRep: 0,
-        targetRep: 1,
-      }
-  ])
+      ...prevTimer.slice(id),
+    ])
+  }
 
   function changeCompletion(id) {
-    setTimers(prevTimer => (
-      [
-        ...prevTimer.slice(0,id - 1),
-        {
-          ...prevTimer[id - 1],
-          isComplete: prevTimer[id - 1].currentRep >= prevTimer[id - 1].targetRep - 1 && true,  // it works due to - 1, but i think there's something wrong with useEffect
-          currentRep: prevTimer[id - 1].currentRep + 1
-        },
-        ...prevTimer.slice(id)
-    ]))
+    const newTimers = timers.map((timer) => ({ ...timer })) // make a deep copy of an array of objects so you don't accidentally update state directly
```

Comment:
I'm wondering about accidentally updating state directly part, I thought that prevTimer was already a copy, so it wouldn't update the state directly unless it was named like the state (timers)?

Reply:
Three things here:
1. I know it may seem small, but it helps with readability and debugging to correctly keep the plurality of variable names. Here, that means that since the state is `timers`, the argument here should be `prevTimers`, not `prevTimer`
2. prevTimers is NOT a copy. It shares the same reference as timers, so if you update prevTimers directly, you are directly mutating state, which should be avoided in React. You can check this yourself like this:
```javascript
  function changeCompletion(id) {
    setTimers((prevTimers) => {
      const newTimers = timers.map((timer) => ({ ...timer })) // make a deep copy of an array of objects
      // when checking equality of arrays & objects with `==` and `===` operators in JS, you are actually checking if their references are the same (read more here https://masteringjs.io/tutorials/fundamentals/compare-arrays)
      console.log({
        hasSameRefs_prevTimers: timers === prevTimers, // true
        hasSameRefs_newTimers: timers === newTimers, // false
      })

    })
  }
```
This logs the following to the console when a timer completes:
![Screen Shot 2022-04-06 at 11 39 22 AM](https://user-images.githubusercontent.com/55759619/162013767-9a8124cf-a4ab-4030-9df2-8a5d1ab42dfe.png)
3. I just did the manipulation outside of the setTimers call for readability, but you could also do this:
```javascript
  function changeCompletion(id) {
    setTimers((prevTimers) => {
      const newTimers = timers.map((timer) => ({ ...timer })) // make a deep copy of an array of objects so you don't accidentally update state directly
      const timerToUpdateIndex = newTimers.findIndex((timer) => timer.id === id)
      const timerToUpdate = newTimers[timerToUpdateIndex]
      const updatedTimer = {
        ...timerToUpdate,
        currentRep: timerToUpdate.currentRep + 1,
        isComplete: timerToUpdate.currentRep + 1 >= timerToUpdate.targetRep,
      }
      newTimers[timerToUpdateIndex] = updatedTimer
      return newTimers
    })
  }
```